Re-format to reduce Docker image file by ~70%

romik-g opened the pull request:
Dockerfile has been re-formatted and image being built to reduce its' size from ~442.2MB to ~188.3MB. Idea is in leaving only one RUN statement, and inside it to perform build, install and cleanup. Change also cleans up build dependencies.

nurupo commented on other/bootstrap_daemon/docker/Dockerfile:
```diff
@@ -11,42 +14,40 @@ RUN apt-get update && apt-get install -y \
+    python3 other/bootstrap_daemon/docker/get-nodes.py >> /etc/tox-bootstrapd.conf && \
+    export AUTO_ADDED_PACKAGES="$(apt-mark showauto)" && \
+    apt-get remove --purge -y $BUILD_PACKAGES $AUTO_ADDED_PACKAGES && \
+    apt-get install -y $RUNTIME_PACKAGES && \
```
Haven't you installed them already on line 21?

romik-g replied:
Yes, that's correct. But libconfig-dev gets removed (from 'apt-mark show auto' and 'apt-get remove --purge') and I get
`/usr/local/bin/tox-bootstrapd: error while loading shared libraries: libconfig.so.9: cannot open shared object file: No such file or directory`
, and libsodium has been added for re-installation just to accompany libconfig.

nurupo commented on other/bootstrap_daemon/docker/Dockerfile:
```diff
@@ -11,42 +14,40 @@ RUN apt-get update && apt-get install -y \
+    N=-1 && \
     while grep -q "bootstrap_nodes =" /etc/tox-bootstrapd.conf; \
     do \
       head -n $N tox-bootstrapd.conf > /etc/tox-bootstrapd.conf; \
```
This should be `other/bootstrap_daemon/tox-bootstrapd.conf` instead of `tox-bootstrapd.conf`, as you have removed `WORKDIR /root/toxcore/other/bootstrap_daemon/`.

romik-g replied:
Correct.

nurupo commented:
It would also help if you don't remove the comments that were in the original Dockerfile.

nurupo commented:
Also, set your git name and email correctly, you have made the commit as `root <root@mds-station.com>`.

romik-g commented:
@nurupo 
> don't remove the comments that were in the original Dockerfile.
Fixed in form of "echo".
> set your git name and email correctly
Fixed.
Thank you.

nurupo commented:
Why `echo`? Don't `# comemnt` comments work?

romik-g commented:
@nurupo just checked, they do work. Strangely they didn't work earlier, could be my mistake.

nurupo commented:
If you do the next changes, then the image size gets further reduced to 130.5 MB, which is just 5.4 MB larger than the Debian Jessie image (125.1 MB). This essentially solves the libconfig error you were getting.

```diff
diff --git a/other/bootstrap_daemon/docker/Dockerfile b/other/bootstrap_daemon/docker/Dockerfile
index bfa8f1e..983ebb4 100644
--- a/other/bootstrap_daemon/docker/Dockerfile
+++ b/other/bootstrap_daemon/docker/Dockerfile
@@ -12,9 +12,10 @@ RUN export BUILD_PACKAGES="\
       git \
       yasm \
       libsodium-dev \
+      libconfig-dev \
       python3" && \
     export RUNTIME_PACKAGES="\
-      libconfig-dev \
+      libconfig9 \
       libsodium13" && \
 # get all deps
     apt-get update && apt-get install -y $BUILD_PACKAGES $RUNTIME_PACKAGES && \
@@ -42,9 +43,9 @@ RUN export BUILD_PACKAGES="\
     done && \
 # add bootstrap nodes from https://nodes.tox.chat/
     python3 other/bootstrap_daemon/docker/get-nodes.py >> /etc/tox-bootstrapd.conf && \
+# perform cleanup
     export AUTO_ADDED_PACKAGES="$(apt-mark showauto)" && \
     apt-get remove --purge -y $BUILD_PACKAGES $AUTO_ADDED_PACKAGES && \
-    apt-get install -y $RUNTIME_PACKAGES && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/* && \
     cd / && \
```

nurupo commented:
And this can help remove 2 intermediate images.

```diff
iff --git a/other/bootstrap_daemon/docker/Dockerfile b/other/bootstrap_daemon/docker/Dockerfile
index 983ebb4..2f1d80f 100644
--- a/other/bootstrap_daemon/docker/Dockerfile
+++ b/other/bootstrap_daemon/docker/Dockerfile
@@ -51,8 +51,6 @@ RUN export BUILD_PACKAGES="\
     cd / && \
     rm -rf /tmp/*
 
-WORKDIR /
-
 USER tox-bootstrapd
 
 ENTRYPOINT /usr/local/bin/tox-bootstrapd \
@@ -60,5 +58,4 @@ ENTRYPOINT /usr/local/bin/tox-bootstrapd \
              --log-backend stdout \
              --foreground
 
-EXPOSE 443 3389 33445
-EXPOSE 33445/udp
+EXPOSE 443/tcp 3389/tcp 33445/tcp 33445/udp
```

nurupo commented:
Though I guess I could make a PR for them myself and not bother you : )

romik-g commented:
libconfig9, oh cool.
WORKDIR and EXPOSE should add 0 MiB to the image size. I would leave WORKDIR as it was before and combine EXPOSE as you did.

nurupo commented:
Yeah, it adds 0 MiB, it's just that it creates new image layers (`docker images -a`), though I guess we don't have to worry about these, they are not even seen unless you use `-a` flag.

romik-g commented:
@nurupo, have a look, please.
I have some other changes prepared for Docker file, but they are up to further discussion and not for this specific Pull request.

nurupo commented:
This looks good. Could you squash your commits into one or two?

nurupo commented:
@GrayHatter this PR is feature complete and I have tested it, so it could be merged as it is, though I'd like the contributor to squash commits into one or two.

GrayHatter commented:
@romik-g once you can squash this down to <4 commits it'll be ready to merge

nurupo commented:
Okay, this wait is ridiculous.
@irungentoo I guess you could squash the commits yourself during the merge, keeping @romik-g's authorship.

GrayHatter commented:
@nurupo, you can merge and squash and open a new PR?
On Feb 25, 2016 2:19 PM, "nurupo" <notifications@github.com> wrote:

> Okay, this wait is ridiculous.
> @irungentoo <https://github.com/irungentoo> I guess you could squash the
> commits yourself during the merge, keeping @romik-g
> <https://github.com/romik-g>'s authorship.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/irungentoo/toxcore/pull/1511#issuecomment-189009753>.
>

nurupo commented:
Why? I will just let @irungentoo do it.

romik-g commented:
Hello. I'm a little bit busy with exams, and new on GitHub. Tell me the command to squash (how many changes?), and I'll do it on weekend or late Friday. Sorry, I have no time now to test squashing, and it could happen that author of patch will be root@mds-station.com.

irungentoo commented:
git rebase -i <commit hash>

is how you rebase.

GrayHatter commented:
@romik-g more info https://git-scm.com/book/en/v2/Git-Branching-Rebasing

GrayHatter commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/irungentoo/toxcore/1511)*
<!-- Sent from Reviewable.io -->

nurupo commented:
@GrayHatter why did you comment this review thing? Just to say that you have reviewed the PR?

romik-g commented:
git rebase -i a47ebbfccd59b22e9796f41b3afdb5cdb218d6be
then there was a menu in editor, I've saved and exited, then commit (result: nothing to commit), and push (result: Everything is up-to-date). So it didn't work.

romik-g commented:
commit_hash was a47ebbfccd59b22e9796f41b3afdb5cdb218d6be

romik-g commented:
Ooh, github shortens long hash to something else - a47ebbf. Don't know how to insert long one.

nurupo commented:
I have already rebate and made #1536

cebe commented:
> Ooh, github shortens long hash to something else - a47ebbf. Don't know how to insert long one.

@romik-g the short hash is equal to the long one, it is unique this way so it is save to refer to a commit.

GrayHatter commented:
@nurupo I was playing with reviewable, it's awesome. But verbose. Didn't mean to make that comment public. I thought it would remain on reviewable

romik-g commented:
@nurupo , @GrayHatter , @irungentoo , do I need to do anything, or I just should leave it like it is in my github account?
I also have some more changes to this docker container description file, bet they are not related to container size. Should I push it to this existing pull request, or I need to create a new one? Thank you!
